Sources/System/Sockets/SocketDescriptor+Messages.swift

```diff
@@ -0,0 +1,539 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension SocketDescriptor {
+  /// A reusable collection of variable-sized ancillary messages
+  /// sent or received over a socket. These represent protocol control
+  /// related messages or other miscellaneous ancillary data.
+  ///
+  /// Corresponds to a buffer of `struct cmsghdr` messages in C, as used
+  /// by `sendmsg` and `recmsg`.
+  public struct AncillaryMessageBuffer {
+    internal var _buffer: _RawBuffer
+    internal var _endOffset: Int
+
+    /// Initialize a new empty ancillary message buffer with no preallocated
+    /// storage.
+    internal init() {
+      _buffer = _RawBuffer()
+      _endOffset = 0
+    }
+
+    /// Initialize a new empty ancillary message buffer of the
+    /// specified minimum capacity (in bytes).
+    internal init(minimumCapacity: Int) {
+      let headerSize = MemoryLayout<CInterop.CMsgHdr>.size
+      let capacity = Swift.max(headerSize + 1, minimumCapacity)
+      _buffer = _RawBuffer(minimumCapacity: capacity)
+      _endOffset = 0
+    }
+
+    internal var _headerSize: Int { MemoryLayout<CInterop.CMsgHdr>.size }
+    internal var _capacity: Int { _buffer.capacity }
+
+    /// Remove all messages currently in this buffer, preserving storage
+    /// capacity.
+    ///
+    /// This invalidates all indices in the collection.
+    ///
+    /// - Complexity: O(1). Does not reallocate the buffer.
+    public mutating func removeAll() {
```

Comment:
Do we want a `keepingCapacity` argument? I guess we probably have a different desired default behavior, but as written a `removeAll()` will keep capacity here but not on Array. I'm assuming we won't make this a RRC, so should we have a different name (e.g. `reset` or `clear`)?